File: ntrp/server/routers/skills.py
from fastapi import APIRouter, Depends, HTTPException

from ntrp.server.runtime import Runtime, get_runtime
from ntrp.server.schemas import InstallRequest
from ntrp.skills.service import SkillService

router = APIRouter(tags=["skills"])


def _require_skill_service(runtime: Runtime = Depends(get_runtime)) -> SkillService:
    if not runtime.skill_service:
        raise HTTPException(status_code=503, detail="Skill service not available")
    return runtime.skill_service


@router.get("/skills")
async def list_skills(svc: SkillService = Depends(_require_skill_service)):
    return {
        "skills": [
            {
                "name": m.name,
                "description": m.description,
                "location": m.location,
            }
            for m in svc.list_all()
        ],
    }


@router.post("/skills/install")
async def install_skill(request: InstallRequest, svc: SkillService = Depends(_require_skill_service)):
    try:
        meta = await svc.install(request.source)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    return {
        "name": meta.name if meta else request.source,
        "description": meta.description if meta else "",
        "status": "installed",
    }


@router.delete("/skills/{name}")
async def remove_skill(name: str, svc: SkillService = Depends(_require_skill_service)):
    if not svc.remove(name):
        raise HTTPException(status_code=404, detail=f"Skill not found: {name}")
    return {"status": "removed", "name": name}

File: ntrp-ui/src/api/skills.ts
import type { Config } from "../types.js";
import { api } from "./fetch.js";

export interface Skill {
  name: string;
  description: string;
  location: string;
}

export async function getSkills(config: Config): Promise<{ skills: Skill[] }> {
  return api.get(`${config.serverUrl}/skills`);
}

export async function installSkill(config: Config, source: string): Promise<{ name: string; description: string; status: string }> {
  return api.post(`${config.serverUrl}/skills/install`, { source }, { timeout: 60000 });
}

export async function removeSkill(config: Config, name: string): Promise<{ status: string }> {
  return api.delete(`${config.serverUrl}/skills/${name}`);
}
